docs/design/data-protection-test.md

```diff
@@ -0,0 +1,219 @@
+# DataProtectionTest CRD and Controller Design
+
+## Abstract
+
+This design introduces the **DataProtectionTest** (DPT) CRD and its controller to evaluate data protection performance—including backup upload and volume snapshot readiness—in an OpenShift cluster. It supports two tests:
+
+1. **Upload Speed Test:** Measures the speed at which a test data is uploaded to cloud object storage using configuration from a BackupStorageLocation.
+2. **CSI VolumeSnapshot Test:** Creates a VolumeSnapshot(s) from a specified PVC source(s) and measures the time taken (as a duration) for the snapshot(s) to become ready.
+
+Additionally, the design includes a mechanism to determine the S3-compatible vendor.
+
+## Background
+
+In several customer environments, backup/restore operations have experienced delays or stalls due to poor network connectivity, throttling at the cloud provider, or misconfigurations in BackupStorageLocation settings. Given the critical role of backup performance in disaster recovery, administrators need real-time, accurate metrics on how fast data uploads occur and how promptly snapshots are created.
+
+This design addresses that need by measuring:
+
+- **Upload speed:** How quickly data is transferred to object storage.
+- **Snapshot performance:** How long a CSI VolumeSnapshot(s) take to become ReadyToUse.
+- **S3 compatibility:** Identifying the specific S3 vendor by inspecting HTTP response headers from the storage endpoint.
+
+## Goals
+
+- Upload data and compute the speed in Mbps.
+- Create a CSI VolumeSnapshot(s) and measure the time taken for it to become ready (reported as a duration).
+- Identify the S3-compatible vendor using an HTTP call that inspects response headers.
+- Expose all results—including upload speed, snapshot ready duration, and S3 vendor—in the CRD status.
+- Must gather will be updated th gather the DPT CR(s) for insights.
+
+
+## Non-Goals
+
+- This design will not create or modify BackupStorageLocation entries in OADP.
+- It will not implement download or latency tests, focusing solely on upload speed.
+- Scheduling of recurring tests is not supported in the initial version.
+- Integration via DPA CR
+
+## High-Level Design
+Components involved and their responsibilities:
+
+- **DataProtectionTest (DPT) CRD:**
+    - **Spec:**
+        - **backupLocation:** Contains Velero-based backup storage configuration.
+        - **uploadSpeedTestConfig:** Test parameters for the upload (file size, test timeout).
+        - **CSIVolumeSnapshotTestConfig:** Test parameters for the CSI VolumeSnapshot test (snapshot class, source PVC name, and importantly the PVC namespace, plus snapshot timeout).
+    - **Status:**
+        - **UploadTestStatus:** Groups upload speed (in Mbps), success flag, and error messages.
+        - **SnapshotTestStatus:** Groups the snapshot test results, reporting status and the duration taken for the snapshot to be ready.
```

Comment:
```suggestion
        - **SnapshotTestStatus:** Groups the snapshot test results, reporting status and the duration taken for the snapshot(s) to be ready.
```

Reply:
Are we reporting per-snapshot duration? That probably makes more sense than total duration, since snapshots could become ready in parallel -- so the cumulative time may be misleading.

Reply:
yeah per snapshot duration.